modules/stochastic_tools/src/userobjects/InverseMapping.C

```diff
@@ -0,0 +1,216 @@
+      _surrogate_models[var_i]->evaluate(_input_parameters, reduced_coefficients);
+    else
+      reduced_coefficients = _input_parameters;
+
+    // The sanity check on the size of reduced_coefficients is happening in the inverse mapping
+    // function
+    DenseVector<Real> reconstructed_solution;
+    _mapping->inverse_map(
+        _var_names_to_reconstruct[var_i], reduced_coefficients, reconstructed_solution);
+
+    MooseVariableFieldBase * var_to_fill = _variable_to_fill[var_i];
+    const MooseVariableFieldBase * var_to_reconstruct = _variable_to_reconstruct[var_i];
+
+    // We set the global DoF indices of the requested variable. The underlying assumption here is
+    // that we are reconstructing the solution in an application which has the same ordering in the
+    // extracted `dofs` vector as the one which was used to serialize the solution vectors in
+    // `SerializedSolutionTransfer`.
+    nl.setVariableGlobalDoFs(_var_names_to_reconstruct[var_i]);
+
+    // Get the DoF indices
+    const auto & dofs = nl.getVariableGlobalDoFs();
+
+    // Get the dof map to be able to determine the local dofs easily
+    const auto & dof_map = var_to_reconstruct->sys().system().get_dof_map();
+    dof_id_type local_dof_begin = dof_map.first_dof();
+    dof_id_type local_dof_end = dof_map.end_dof();
+
+    // Populate the temporary vector with the reconstructed solution
+    for (const auto & dof_i : index_range(dofs))
+      if (dofs[dof_i] >= local_dof_begin && dofs[dof_i] < local_dof_end)
+        temporary_vector->set(dofs[dof_i], reconstructed_solution(dof_i));
+
+    // Get the system and variable numbers for the dof objects
+    unsigned int to_sys_num = _variable_to_fill[var_i]->sys().system().number();
+    unsigned int to_var_num =
+        var_to_fill->sys().system().variable_number(_var_names_to_fill[var_i]);
+
+    unsigned int from_sys_num = var_to_reconstruct->sys().system().number();
+    unsigned int from_var_num =
+        var_to_reconstruct->sys().system().variable_number(_var_names_to_reconstruct[var_i]);
+
+    // Get a link to the mesh for the loops over dof objects
+    const MeshBase & to_mesh = _fe_problem.mesh().getMesh();
+
+    // If the variable is nodal, we can do a nodal loop
+    if (_is_nodal[var_i])
+    {
+      for (const auto & node : to_mesh.local_node_ptr_range())
+      {
+        // We have nothing to do if we don't have dofs at this node
+        if (node->n_dofs(to_sys_num, to_var_num) < 1)
+          continue;
+
+        // Get the dof ids for the from/to variables
+        const auto & to_dof_id = node->dof_number(to_sys_num, to_var_num, 0);
+        const auto & from_dof_id = node->dof_number(from_sys_num, from_var_num, 0);
+
+        // Fill the dof of the auxiliary variable using the dof of the temporary variable
+        var_to_fill->sys().solution().set(to_dof_id, (*temporary_vector)(from_dof_id));
+      }
+    }
+    // Otherwise we have to do an element loop
```

Comment:
what if we have both nodal and elemental dofs

Reply:
Could you show me an example in moose where one variable has both types of dofs? I looked at most of our transfers and they operate with the same logic. 
https://github.com/idaholab/moose/blob/next/framework/src/transfers/MultiAppShapeEvaluationTransfer.C

Reply:
copy transfer does it for both

Unfortunately they would be exotic and I dont use those.
@roystgnr what's a variable that has both nodal & elemental dofs

Reply:
Skimming through, I see: `BERNSTEIN`/`RATIONAL_BERNSTEIN` on `TRI6` for p>2, `HERMITE` on `QUAD8`/`QUAD9`/`HEX20`/`HEX27` for p>3, `HIERARCHIC` on `EDGE2`/`EDGE3`/`QUAD8`/`QUAD9`/`HEX27` for p>1 or `TRI6` for p>2, `SZABAB` (Szabo-Babushka) on `EDGE2` for p>2 or `TRI6`/`QUAD8` for p>3.

Not sure if there's anything in there that MOOSE users currently use regularly, but there's a bit of a vicious cycle between "some MOOSE objects break when you try to do more than the most basic FE spaces" versus "MOOSE users learn subconsciously to avoid what should be reasonable ways to improve solution convergence" that I would like to see us break out of.

Reply:
Thank you!

Reply:
I agree with the conclusion. Peter raised a good point that some transfers dont do it well